File: interview/questions/Fibonacci.java
package leetcode.interview.questions;

public class Fibonacci<E> {

	// Solution 1: Iterative
	// Time complexity: O(n)
	// Space complexity: O(1)
	public static int Fibonacci(int n) {
		if (n <= 1)
			return n;

		int sum, first = 0, second = 1;

		while (n-- > 1) {
			sum = first + second;
			first = second;
			second = sum;
		}
		return second;
	}

	// Solution 2: Recursive
	// Time complexity: O(2^n)
	// T(n) = T(n-1) + T(n-2)is an exponential time
	// Space complexity: O(n)
	// space for recursive function call stack
    public static int fib(int N)
    {
        if(N == 0) return 0;
        else if(N <= 2)
            return 1;
        else
            return fib(N - 1) + fib(N - 2);
    }

    
	public static int countWays(int n) {
		if (n < 0) {
			return 0;
		} else if (n == 0) {
			return 1;
		} else {
			return countWays(n - 1) + countWays(n - 2);
		}
	}
	
	
	public static void main(String[] args) {

		System.out.println(fib(8));
		
		int n = 5;
		int ways = countWays(n);
		System.out.println(ways);

	}
}

File: interview/questions/easy/array/fibonacci .js
/**
 * @param {number} n
 * @return {number}
 */
var fib = function(n) {
    if (n < 2) return n;
    
    let a = 0;
    let b = 1;
    let sum = a + b;

    while(n>1){
        sum= a + b;
        a= b;
        b= sum;
        n--;
    }
    return sum;
}

console.log(fib(8))
console.log(fib(5))
console.log(fib(9))
console.log(fib(10))

// /**
//  * @param {number} n
//  * @return {number}
//  */
// const fib = function(n) {
//     return n<2 ? n : fib(n-1) + fib(n-2);
// };

// console.log(fib(0))
// console.log(fib(1))
// console.log(fib(2))
// console.log(fib(3))